### PR TITLE
define Yapf config

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

eventually, we would move to Yet-Another-Python-Formater so this is just setting rules, for illustration how it goes you can check Bolts codebase when we have already deployed it
With deploy YAPF check we may drop ISORT which has atm incompatibility with PEP8, see https://github.com/PyCQA/isort/issues/1594#issuecomment-763183604
discussed with @tchaton and finetuned config with @akihironitta 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
So why yapf over alternatives? (black, for example: https://news.ycombinator.com/item?id=17155048)

**Comment by Borda** [comment]
> So why yapf over alternatives? (black, for example: https://news.ycombinator.com/item?id=17155048)

We have already spoke about it, black is just one (no way to configure as far as I know) so you either like it or hate it... 
On the other hand with yapf you have some freedom to set it up...

**Comment by carmocca** [comment]
> And my experience is that black is less deterministic and much longer which makes it more difficult to read...

Never used yapf but reading online, seems like its the other way around.

What I care about the most is that the formatter only allows one consistent single style. If yapf does, then that's OK with me

**Comment by awaelchli** [comment]
the drawback is that if you change your mind about the config, you have to reformat the whole code base. 
I have one wish, I beg you, please if it's possible can we avoid this code style:

```python
foo_long_function_(arg1, arg2, arg3
                   arg4, arg5, arg6)
```
it leads to uneven indentation which is just horrible in my opinion.

**Comment by Borda** [comment]
> I have one wish, I beg you, please if it's possible can we avoid this code style:
> 
> ```python
> foo_long_function_(arg1, arg2, arg3
>                    arg4, arg5, arg6)
> ```

Heh, that you are lucky guy, I hate it too =D

**Review by tchaton — APPROVED** [review]
Love it ! Great addition !

**Review by ydcjeff — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Great addition !

**Comment by akihironitta** [comment]
In Bolts, I had a collision between flake8 and yapf. Applying yapf caused the error `[W503] line break before binary operator` from flake8, so shall we modify the flake8 config to ignore `W503`? https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/462/files#r561509665

**Comment by Borda** [comment]
> In Bolts, I had a collision between flake8 and yapf. Applying yapf caused the error `[W503] line break before binary operator` from flake8, so shall we modify the flake8 config to ignore `W503`? https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/462/files#r561509665

we can configure this in yapf, right?

**Comment by akihironitta** [comment]
@Borda Could you have a look at https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/539? I'm having difficulty in resolving the conflicts between pep8 (flake8) and yapf. It seems that yapf is not really compliant with pep8 even if we set:
```
[yapf]
based_on_style = pep8
```

**Comment by akihironitta** [comment]
In Bolts, we had conflicts between yapf and flake8, and @Borda resolved errors from flake8 by writing those lines in a different way. https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/539 So, here, I don't think we need to modify our flake8 config for now.

**Comment by Borda** [comment]
I still think that you shall write reasonable code, and there are always multiple ways... :]

**Review by williamFalcon — APPROVED** [review]

File: setup.py [hunk 1]
```diff
@@ -46,9 +46,7 @@
 extras['all'] = extras['dev'] + extras['examples']  # + extras['docs']
 
 # These packages shall be installed only on GPU machines
-PACKAGES_GPU_ONLY = (
-    'horovod',
-)
+PACKAGES_GPU_ONLY = ('horovod')
```
Comment (awaelchli) [inline setup.py:49]: what the heck?
that's not equivalent
one is a tuple, the other is a string!
if yapf did this automatically, that's highly concerning

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: setup.py (hunk 1 above)
Comment (Borda) [reply setup.py:49]: heh, that was my bad, I remove the, manually :/

Comment (awaelchli) [reply setup.py]: ok, it looks like it's not yapf, just a manual edit

**Review by awaelchli — APPROVED** [review]

File: .yapfignore [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# TODO
+pl_examples/*
+
+# TODO
+pytorch_lightning/*
+
+# TODO
+tests/*
```
Comment (awaelchli) [inline .yapfignore:8]:
```suggestion
.git/*

# TODO
pl_examples/*

# TODO
pytorch_lightning/*

# TODO
tests/*
```
it seems this is needed on windows, without it running yapf on root folder crashes on python files in .git folder
